### ISSUE TITLE
M3 Fusion

### ISSUE THREAD

**Issue by FerdinandKlingenberg** [remicres/otbtf#70]
Dear Remi, I got a question about the M3 Fusion model, which is provided in the python example folder.

I run the M3Fusion model on a time-series dataset consisting of 4 different Sentinel-2 cloud-free mosaic products (10 bands each mosaic, all bands resampled to 10m), and for the VHR, I use some 4 bands Pleiades data (2m resolution) over the same area. I try to classify some roads (only one class is used for training)
I try your setup on the example page, but I fail somehow: 
As an example of what I do, the one S2 patch follows like this:

"#" Patches extraction A
otbcli_PatchesExtraction \\
-source1.il "/home/otbuser/2021_mosaic_norm.tif" \\
-source1.patchsizex 1 \\
-source1.patchsizey 1 \\
-source1.out "/home/otbuser/2021_mosaic_norm_patches_A.tif" \\
-source2.il "/home/otbuser/2021_mosaic_norm_LabelImage.tif" \\
-source2.patchsizex 1 \\
-source2.patchsizey 1 \\
-source2.out "/home/otbuser/2021_mosaic_norm_Bitmap_patches_A.tif" \\
-vec "/home/otbuser/roaddata/Centroids/Centroid_A.shp" \\
-field "id" 

...
Then the same goes for the B validation variant and the other three cloud-free mosaics. 
The VHR dataset is very similar, whereas the essential change is the patch size of 25

"#" Patches extraction A
otbcli_PatchesExtraction \\
-source1.il "/home/otbuser/VHR_norm.tif" \\
-source1.patchsizex 25 \\
-source1.patchsizey 25 \\
-source1.out "/home/otbuser/VHR_norm_patches_A.tif" \\
-source2.il "/home/otbuser/VHR_norm_LabelImage.tif" \\
-source2.patchsizex 25 \\
-source2.patchsizey 25 \\
-source2.out "/home/otbuser/VHR_norm_Bitmap_patches_A.tif" \\
-vec "/home/otbuser/roaddata/Centroids/Centroid_A.shp" \\
-field "id" 
...
And then the same for the B validation variant. 

The M3 Fusion model I create with:
python create_savedmodel_ienco-m3_patchbased.py --outdir /home/otbuser/M3Fusion_model \\
--n_timestamps 4 --n_dims 10 --nclasses 1

The other adjustable settings should follow the default ones, right?

Then for the training:
export OTB_TF_NSOURCES=2

otbcli_TensorflowModelTrain \\
-model.dir "/home/otbuser/M3Fusion_model" \\
-model.saveto "/home/otbuser/M3Fusion_model/variables/variables" \\
-training.source1.il /home/otbuser/2021_mosaic_norm_patches_A.tif ..._A.tif ..._A.tif ..._A.tif \\
-training.source1.patchsizex 1 -training.source1.patchsizey 1 -training.source1.placeholder "x_rnn" \\
-training.source2.il /home/otbuser/VHR_norm_patches_A.tif \\
-training.source2.patchsizex 25 -training.source2.patchsizey 25 -training.source2.placeholder "x_cnn" \\
-training.source3.il /home/otbuser/2021_mosaic_norm_Bitmap_patches_A.tif ..._A.tif ..._A.tif ..._A.tif \\
-training.source3.patchsizex 1 -training.source3.patchsizey 1 -training.source3.placeholder "y" \\
-training.targetnodes "optimizer" \\
-training.userplaceholders "is_training=true" "drop_rate=0.1" "learning_rate=0.0002" \\
-validation.mode "class" -validation.step 1 \\
-validation.source1.il S2..._B.tif ..._B.tif ..._B.tif ..._B.tif \\
-validation.source1.name "x_rnn" \\
-validation.source2.il /home/otbuser/VHR_norm_Bitmap_patches_B.tif \\
-validation.source2.name "x_cnn" \\
-validation.source3.il S2..._B.tif ..._B.tif ..._B.tif ..._B.tif \\
-validation.source3.name "prediction"

The error message follows: 
Training epoch "#"1: 100% [**************************************************]2022-01-13 19:12:57 (FATAL) TensorflowModelTrain: itk::ERROR: TensorflowMultisourceModelTrain(0x5565d767eb20): Can't run the tensorflow session !
Tensorflow error message:
Invalid argument: logits and labels must have the same first dimension, got logits shape [100,1] and labels shape [400]
         [[{{node sparse_softmax_cross_entropy_loss_2/xentropy/xentropy}}]]
OTB Filter debug message:
Output image buffered region: ImageRegion (0x7ffea2859d30)
  Dimension: 2
  Index: [0, 0]
  Size: [1, 1]

Input "#"0:
Requested region: ImageRegion (0x7ffea2859d60)
  Dimension: 2
  Index: [0, 0]
  Size: [1, 23324]

Tensor shape ("x_rnn": {100, 1, 1, 40}
Input "#"1:
Requested region: ImageRegion (0x7ffea2859d60)
  Dimension: 2
  Index: [0, 0]
  Size: [25, 583100]

Tensor shape ("x_cnn": {100, 25, 25, 4}
Input "#"2:
Requested region: ImageRegion (0x7ffea2859d60)
  Dimension: 2
  Index: [0, 0]
  Size: [1, 23324]

Tensor shape ("y": {100, 1, 1, 4}
User placeholders:
is_training Tensor shape is {0} data type is 10

drop_rate Tensor shape is {0} data type is 1

learning_rate Tensor shape is {0} data type is 1

exit status 1

Any help to solve this problem is much appreciated 
Thanks,
Ferdinand Klingenberg

**Comment by remicres** [comment]
Hi @FerdinandKlingenberg 

Long time I didn't use this model... I hope it still works.
In:
```
python create_savedmodel_ienco-m3_patchbased.py --outdir /home/otbuser/M3Fusion_model \
--n_timestamps 4 --n_dims 10 --nclasses 1
```
The model should have at least 2 classes in order to build a valid softmax function.
Anyway, it looks like there is a bug somewhere, because the loss function works on labels of different sizes (1=the number of class you have specified VS 4=where comes this number?)

it looks like here:
```
...
-validation.source3.il S2..._B.tif ..._B.tif ..._B.tif ..._B.tif \
-validation.source3.name "prediction"
...
```
You put an image which has 4 components. However this image should be the label image (i.e. a monoband integer labels image)

**Comment by FerdinandKlingenberg** [comment]
Hi @remicres,

Thanks a lot! Removing three out of four label files and changing the _nclasses_ to 2 worked! 
Thanks for this quick fix 👍 

For the testing purpose on the inference step, I run the resulting model on the same four Sentinel-2 images and the one VHR image I used for training, with the following command:
 
otbcli_TensorflowModelServe \\
-source1.il /home/otbuser/2021_mosaic_norm.tif \\
..._1.tif \\
..._2.tif \\
..._3.tif \\
-source1.rfieldx 1 -source1.rfieldy 1 -source1.placeholder "x_rnn" \\
-source2.il /home/otbuser/VHR_norm.tif \\
-source2.rfieldx 25 -source2.rfieldy 25 -source2.placeholder "x_cnn" \\
-model.dir "/home/otbuser/M3Fusion_model" \\
-output.names "prediction" -out "/home/otbuser/M3Fusion.tif" uint8

and I receive this error:
2022-01-14 13:38:10 (FATAL) TensorflowModelServe: itk::ERROR: Input image size number 1 mismatch
exit status 1

My temporary workaround is to stack all the four images to one single file, in other words, one file with 40 bands. The create  
model is similar apart from _--n_timestamps 1_ _--nclasses 2_ and the ModelTraining and ModelServe are working fine when the time-series dataset is one stacked file (patches and labels are also changed accordingly). Why it is like this, I don't know. Have you seen this error message before, and will one large stacked (40 bands) file produce the same as delivering 4 files with 10 bands each when using the M3Fusion?

Ferdinand

**Comment by remicres** [comment]
Hi @FerdinandKlingenberg 

It looks like your `..._1.tif ..._2.tif ..._3.tif ` have not the same size.
Maybe the tool used for the layer stack has performed a resampling or a cropping of the images?

**Comment by FerdinandKlingenberg** [comment]
Hi @remicres,

You are right! It worked as intended after some finetuning of the image size and corner coordinates. Thanks a lot for your help!
